frontend/src/metabase/lib/expressions/recursive-parser.js

```diff
@@ -0,0 +1,274 @@
+      const { op } = next();
+      expr = [op, expr, parseAdditive()];
+    }
+    return expr;
+  };
+
+  const parseBooleanUnary = () => {
+    if (matchOps([OP.Not])) {
+      const { op } = next();
+      return [op, parseBooleanUnary()];
+    }
+    return parseComparison();
+  };
+
+  const parseBooleanAnd = () => {
+    let expr = parseBooleanUnary();
+    while (matchOps([OP.And])) {
+      const { op } = next();
+      if (Array.isArray(expr) && expr[0] === op) {
+        expr.push(parseBooleanUnary());
+      } else {
+        expr = [op, expr, parseBooleanUnary()];
+      }
+    }
+    return expr;
+  };
+
+  const parseBooleanOr = () => {
+    let expr = parseBooleanAnd();
+    while (matchOps([OP.Or])) {
+      const { op } = next();
+      if (Array.isArray(expr) && expr[0] === op) {
+        expr.push(parseBooleanAnd());
+      } else {
+        expr = [op, expr, parseBooleanAnd()];
+      }
+    }
+    return expr;
+  };
+
+  const booleanExpression = () => parseBooleanOr();
+
+  const parseExpression = () => booleanExpression();
+
+  return parseExpression();
+}
+
+const modify = (node, transform) => {
+  if (Array.isArray(node)) {
+    const [operator, ...operands] = node;
+    return transform([
+      operator,
+      ...operands.map(sub => modify(sub, transform)),
+    ]);
+  }
+  return transform(node);
+};
+
+const NEGATIVE_FILTER_SHORTHANDS = {
+  contains: "does-not-contain",
+  "is-null": "not-null",
+  "is-empty": "not-empty",
+};
+
+// ["NOT", ["is-null", 42]] becomes ["not-null",42]
+const useShorthands = tree =>
+  modify(tree, node => {
+    if (Array.isArray(node) && node.length === 2) {
+      const [operator, operand] = node;
+      if (operator === OP.Not && Array.isArray(operand)) {
+        const [fn, ...params] = operand;
+        const shorthand = NEGATIVE_FILTER_SHORTHANDS[fn];
+        if (shorthand) {
+          return [shorthand, ...params];
+        }
+      }
+    }
+    return node;
+  });
+
+const adjustOptions = tree =>
+  modify(tree, node => {
+    if (Array.isArray(node)) {
+      const [operator, ...operands] = node;
+      if (operands.length > 0) {
+        const clause = MBQL_CLAUSES[operator];
+        if (clause && clause.hasOptions) {
+          if (operands.length === clause.args.length + 1) {
+            // the last one holds the function options
+            const options = operands[operands.length - 1];
+
+            // HACK: very specific to some string/time functions for now
+            if (options === "case-insensitive") {
+              operands.pop();
+              operands.push({ "case-sensitive": false });
+            } else if (options === "include-current") {
+              operands.pop();
+              operands.push({ "include-current": true });
+            }
+            return [operator, ...operands];
+          }
+        }
+      }
+    }
+    return node;
+  });
+
+// ["case", X, Y, Z] becomes ["case", [[X, Y]], { default: Z }]
+const adjustCase = tree =>
```

Comment:
This is broken out into `compiler_passes.ts`, maybe move this function there so pratt parser can use it?

Reply:
These are now exported. We can decide whether to separate them off to a different file if necessary, once the Pratt parser is in place.